MGMT-7153 Support multiplatform test and starting vsphere platform basic flow

michaellevy101 opened the pull request.

eliorerz commented:
/hold 
/cc @YuviGold @lalon4 @nshidlin @osherdp 
/uncc @romfreiman

eliorerz commented:
@michaellevy101 Add description on the PR

eliorerz commented on discovery-infra/test_infra/controllers/node_controllers/libvirt_controller.py:
```diff
@@ -22,9 +24,10 @@
     TEST_DISKS_PREFIX = "ua-TestInfraDisk"
 
-    def __init__(self, private_ssh_key_path: Path):
+    def __init__(self, config: ControllerConfig, cluster_config: BaseClusterConfig):
```
You changed the signature but some of usages left unchanged  e.g. `download_logs.py`

eliorerz commented on discovery-infra/test_infra/helper_classes/config/controller_config.py:
```diff
@@ -0,0 +1,33 @@
+    is_ipv6: bool = None   # Todo - Might change during MGMT-5370
+    bootstrap_in_place: bool = None
+    private_ssh_key_path: Path = None
+    working_dir: str = consts.WORKING_DIR
```
Set to `None` and get set the value on `global_variables` object

michaellevy101 replied:
why? the work dir is constant not an env variable

eliorerz commented on discovery-infra/test_infra/helper_classes/config/controller_config.py:
```diff
@@ -0,0 +1,33 @@
+
+
+@dataclass
+class ControllerConfig(_BaseConfig, ABC):
```
Rename to `BaseControllerConfig` to keep with the conventions

YuviGold replied:
The name should be nodes config, no? This is not just a generic config that would be used from any future controller.

michaellevy101 replied:
It's a common controller configurations not a node configuration

eliorerz commented on discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py:
```diff
@@ -0,0 +1,136 @@
+        self._tf.destroy()
+
+    def shutdown_node(self, node_name: str) -> None:
+        pass
```
Raise `NotImplementedError`

osherdp replied:
```suggestion
        raise NotImplementedError
```
and the same for all other non implemented methods

eliorerz commented on discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py:
```diff
@@ -0,0 +1,136 @@
+        pass
+
+    def shutdown_all_nodes(self) -> None:
+        pass
```
Raise `NotImplementedError`

eliorerz commented on discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py:
```diff
@@ -0,0 +1,136 @@
+        pass
+
+    def start_node(self, node_name: str, check_ips: bool) -> None:
+        pass
```
Same here, and for all unimplemented abstract methods

eliorerz commented on discovery-infra/test_infra/helper_classes/cluster.py:
```diff
@@ -660,8 +660,7 @@ def _set_dns(self, api_vip, ingress_vip):
     def prepare_for_installation(self, static_network_config=None, **kwargs):
         self.update_config(**kwargs)
-        log.info(f"Preparing for installation with clsuter configurations: cluster_config={self._config}")
-        log.info(f"Preparing for installation with nodes configurations: nodes_config={self.nodes.config}")
```
Why not logging it?

eliorerz commented on discovery-infra/test_infra/helper_classes/nodes.py:
```diff
@@ -21,16 +21,11 @@ def __init__(self, node, cluster_host):
     DEFAULT_STATIC_IP_CONFIG = False
 
-    def __init__(self, node_controller: NodeController, private_ssh_key_path: Path):
+    def __init__(self, node_controller: NodeController):
```
This one breaks `test_scaleup` 
FYI @lalon4 @nshidlin

osherdp replied:
it also breaks our gathering of sos-report:
https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_assisted-test-infra/949/pull-ci-openshift-assisted-test-infra-master-e2e-metal-assisted/1414354554329239552/artifacts/e2e-metal-assisted/baremetalds-assisted-gather/build-log.txt

eliorerz commented on discovery-infra/test_infra/tools/terraform_utils.py:
```diff
@@ -20,13 +23,32 @@ def __init__(self, working_dir: str):
     def init_tf(self) -> None:
         self.tf.cmd("init -plugin-dir=/root/.terraform.d/plugins/", raise_on_error=True)
 
+    def select_defined_variables(self, **kwargs):
```
Document what `kwargs` are in this function

eliorerz commented on discovery-infra/test_infra/tools/terraform_utils.py:
```diff
@@ -20,13 +23,32 @@ def __init__(self, working_dir: str):
             logging.error(message)
             raise Exception(message)
 
+    def set_and_apply(self, refresh: bool = True, **kwargs) -> None:
```
Document what `kwargs` are in this function

eliorerz commented on discovery-infra/test_infra/utils/terraform_util.py:
```diff
@@ -0,0 +1,26 @@
+
+    @classmethod
+    def get_folder(cls, cluster_name: str, namespace=None):
+        folder_name = f'{cluster_name}__{namespace}' if namespace else f'{cluster_name}'
```
Double quotes

eliorerz commented on discovery-infra/tests/base_test.py:
```diff
@@ -34,29 +38,105 @@
+
+    @pytest.fixture
+    def nodes(self, controller):
+        return Nodes(controller)
```
I don't think we need it in separate fixture

eliorerz commented on discovery-infra/tests/base_test.py:
```diff
@@ -34,29 +38,105 @@
+        return TerraformConfig()
+
+    @pytest.fixture
+    def prepare_controller_configuration(self, controller_configuration):
```
Add type hint and return hint annotations

eliorerz commented on discovery-infra/tests/base_test.py:
```diff
@@ -34,29 +38,105 @@
+        return Nodes(controller)
+
+    @pytest.fixture
+    def prepare_nodes(self, nodes, cluster_configuration):
```
Add type hint and return hint annotations

eliorerz commented on discovery-infra/tests/base_test.py:
```diff
@@ -34,29 +38,105 @@
+        return ClusterConfig()
+
+    @pytest.fixture
+    def controller(self, cluster_configuration, prepare_controller_configuration):
```
Add type hint and return hint annotations

eliorerz commented on discovery-infra/tests/base_test.py:
```diff
@@ -34,29 +38,105 @@
+        return TerraformController(prepare_controller_configuration, cluster_config=cluster_configuration)
+
+    @pytest.fixture
+    def nodes(self, controller):
```
Add type hint and return hint annotations

eliorerz commented on discovery-infra/tests/base_test.py:
```diff
@@ -34,29 +38,105 @@
+            infra_utils.run_command(f"rm -f {cluster_configuration.iso_download_path}", shell=True)
+
+    @pytest.fixture
+    def prepare_network(self, prepare_nodes, prepare_controller_configuration):
```
Add type hint and return hint annotations

eliorerz commented on discovery-infra/tests/test_e2e_install.py:
```diff
@@ -1,32 +1,40 @@
 class TestInstall(BaseTest):
 
+    @pytest.fixture
+    def cluster_configuration(self, request):
```
Why? 
Just set it on the test itself. One line vs ten...

michaellevy101 replied:
to allow you to override it...

eliorerz reviewed (commented):
I think this implementation (`BaseTest` implementation) give us more building blocks (fixtures to use) but add more and unneeded complexity.
Also, the `get_nodes` flow was left unchanged so it is basically duplicate flow of the current one.

osherdp commented on discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py:
```diff
@@ -0,0 +1,136 @@
+        vms_object_type = self._tf.get_resources(resource_type="vsphere_virtual_machine")
+
+        if vms_object_type:
+            return list()
```
```suggestion
            return []
```

osherdp commented on discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py:
```diff
@@ -0,0 +1,136 @@
+            return Node(name=terraform_vm_state["name"], private_ssh_key_path=self._config.private_ssh_key_path,
+                        node_controller=self)
+
+        return list(map(vsphere_vm_to_node, vms))
```
```suggestion
        return [vsphere_vm_to_node(vm) for vm in vms]
```

osherdp commented on discovery-infra/test_infra/controllers/node_controllers/vsphere_controller.py:
```diff
@@ -0,0 +1,136 @@
+
+        vms = vms_object_type[0].instances
+
+        def vsphere_vm_to_node(terraform_vm_state):
+            return Node(name=terraform_vm_state["name"], private_ssh_key_path=self._config.private_ssh_key_path,
+                        node_controller=self)
```
I usually prefer those to be private ("helper") methods of the class, but that's not a blocker imho

YuviGold commented on terraform_files/vsphere/main.tf:
```diff
@@ -0,0 +1,93 @@
+provider "vsphere" {
+  user                 = var.vsphere_username
```
Elaborate on the description of each data section.
In addition, I'd be glad if you could add a README providing an overview about using `vsphere` provider.

michaellevy101 replied:
It fully documented on the terraform provider why to document it here? I have to document the variables i have created or special behavior

YuviGold commented on terraform_files/vsphere/main.tf:
```diff
@@ -0,0 +1,93 @@
+  datacenter       = var.vsphere_datacenter
+  datastore        = var.vsphere_datastore
+  source_file      = var.iso_download_path
+  destination_file = "ISOs/${basename(var.iso_download_path)}"
```
The `iso_download_path` should be absolute

michaellevy101 replied:
This property is the path on the vsphere itself

osherdp commented on discovery-infra/test_infra/helper_classes/config/controller_config.py:
```diff
@@ -0,0 +1,33 @@
+    worker_disk: int = None
+    worker_disk_count: int = None
+
+    network_mtu: int = None
```
configure your IDE to add new lines on Python modules :)

YuviGold commented on terraform_files/vsphere/main.tf:
```diff
@@ -0,0 +1,93 @@
+  folder               = var.cluster_name
+  enable_disk_uuid     = "true"
+  wait_for_guest_net_timeout  = "0"
+  wait_for_guest_net_routable = "false"
```
Why not?

YuviGold commented on terraform_files/vsphere/main.tf:
```diff
@@ -0,0 +1,93 @@
+  disk {
+    label            = "disk0"
+    size             = var.master_disk_size_gib
+    eagerly_scrub    = false
+    thin_provisioned = true
```
What does it mean? please add a description.

michaellevy101 replied:
https://registry.terraform.io/providers/hashicorp/vsphere/latest/docs/resources/virtual_machine#eagerly_scrub
https://registry.terraform.io/providers/hashicorp/vsphere/latest/docs/resources/virtual_machine#thin_provisioned

osherdp commented on discovery-infra/test_infra/helper_classes/config/controller_config.py:
```diff
@@ -0,0 +1,33 @@
+    master_vcpu: int = None
+    masters_count: int = None
+    master_cpu_mode: str = None
+    master_disk: int = None   # disk size in MB.
+    master_disk_size_gib: str = None # disk size in GB.
```
it doesn't sound fun to have two sources of truth
maybe save the int one, and have a small function that returns the other value?

YuviGold commented on terraform_files/vsphere/main.tf:
```diff
@@ -0,0 +1,93 @@
+    network_id = data.vsphere_network.network.id
+  }
+  
+  disk {
```
Should follow naming like we have for libvirt
https://github.com/openshift/assisted-test-infra/blob/51d5d1c28822dba48ac56dd4210117d0b606a272/terraform_files/baremetal/main.tf#L100-L107

YuviGold commented on discovery-infra/tests/test_e2e_install.py:
```diff
@@ -1,32 +1,40 @@
-    def test_dummy(self, configs: Tuple[ClusterConfig, TerraformConfig], get_nodes, get_cluster, sleep_time):
-        cluster_config, tf_config = configs
-        new_cluster = get_cluster(get_nodes(tf_config, cluster_config), cluster_config)
-        new_cluster.prepare_for_installation()
+    def test_dummy(self, cluster, sleep_time):
+        cluster.prepare_for_installation()
         time.sleep(sleep_time)
```
What's this test? @eliorerz :)

michaellevy101 replied:
irrelevant,  please open a new task for this test...

YuviGold commented on discovery-infra/tests/test_e2e_install.py:
```diff
@@ -1,32 +1,40 @@
 class TestInstall(BaseTest):
 
+    @pytest.fixture
+    def cluster_configuration(self, request):
```
Who calls it?

michaellevy101 replied:
It's a fixture override,  I will document it

osherdp commented on discovery-infra/test_infra/tools/terraform_utils.py:
```diff
@@ -20,13 +23,32 @@ def __init__(self, working_dir: str):
+        supported_variables = self.get_variable_list()
+        return {k: v for k, v in kwargs.items() if k in supported_variables}
+
+    def get_variable_list(self):
```
```suggestion
    def get_variables(self):
```

YuviGold commented on discovery-infra/test_infra/controllers/node_controllers/node_controller.py:
```diff
@@ -3,12 +3,18 @@
 
+    def __init__(self, config: ControllerConfig, cluster_config: BaseClusterConfig):
+        self._config = config
+        self._cluster_config = cluster_config
```
Why do the nodes needs the cluster config?

michaellevy101 replied:
It needs the cluster name and maybe more, for the extended controllers

YuviGold commented on discovery-infra/test_infra/helper_classes/config/nodes_config.py:
```diff
@@ -37,14 +22,10 @@ class BaseTerraformConfig(_BaseConfig, ABC):
     libvirt_worker_ips: List[str] = None
```
so this variable of `libvirt_worker_ips` shouldn't be under `TFConfig` or the name should be `LibvirtConfig`

michaellevy101 replied:
can't fix everything in one PR,  keep the original code ( If i remember correctly)

osherdp commented on discovery-infra/test_infra/tools/terraform_utils.py:
```diff
@@ -36,10 +58,14 @@ def change_variables(self, variables: Dict[str, str], refresh: bool = True) -> N
 
+    def get_resources(self, resource_type: str = None) -> list:
+        state = self.get_state()
+        return [resource for resource in state.resource if resource_type is None or resource["type"] == resource_type]
```
```suggestion
        return [resource for resource in state.resource
                if resource_type is None or resource["type"] == resource_type]
```
just a nit

osherdp commented on discovery-infra/test_infra/utils/terraform_util.py:
```diff
@@ -0,0 +1,26 @@
+from test_infra.utils import utils
+
+
+class TerraformControllerUtil:
```
maybe just delete this class and put all those methods in the module?
the module itself is named ``terraform_util``

osherdp commented on discovery-infra/test_infra/utils/terraform_util.py:
```diff
@@ -0,0 +1,26 @@
+import logging
```
rename module to ``terraform_utils.py``

YuviGold reviewed (commented):
Semi review

Please split this PR into two PRs
* Generic infra changes
* VSphere controller addition

In addition, please provide a very detailed description of the PRs. It cannot be reviewed otherwise.

osherdp commented on terraform_files/vsphere/terraform.tfvars.json:
```diff
@@ -0,0 +1,4 @@
+{
```
this is an example file?
if so I would add more examples for values that usually we don't know how they should look (usually addresses/server urls that can be constructed in many ways)

osherdp commented on discovery-infra/tests/base_test.py:
```diff
@@ -34,29 +38,105 @@
+
+    @pytest.fixture
+    def prepare_network(self, prepare_nodes, prepare_controller_configuration):
+        if global_variables.platform != consts.Platforms.BARE_METAL and global_variables.platform != consts.Platforms.NONE:
```
```suggestion
        if global_variables.platform not in (consts.Platforms.BARE_METAL, consts.Platforms.NONE):
```

osherdp commented on discovery-infra/tests/base_test.py:
```diff
@@ -34,29 +38,105 @@
+            nat.remove_nat_rules()
+
+    @pytest.fixture
+    def cluster(self, api_client, request, proxy_server, prepare_network, cluster_configuration):
+        logging.debug(f'--- SETUP --- Creating cluster for test: {request.node.name}\n')
+        cluster = Cluster(api_client=api_client, config=cluster_configuration, nodes=prepare_network)
+        if cluster_configuration.is_ipv6:
+            self._set_up_proxy_server(cluster, cluster_configuration, proxy_server)
+        yield cluster
+        if request.node.result_call.failed:
+            logging.info(f'--- TEARDOWN --- Collecting Logs for test: {request.node.name}\n')
+            self.collect_test_logs(cluster, api_client, request.node, cluster.nodes)
+            if global_variables.test_teardown:
+                if cluster.is_installing() or cluster.is_finalizing():
+                    cluster.cancel_install()
+                with suppress(ApiException):
+                    logging.info(f'--- TEARDOWN --- deleting created cluster {cluster.id}\n')
+                    cluster.delete()
+
```
this part needs a bit more spacing to help readability

osherdp commented on terraform_files/vsphere/main.tf:
```diff
@@ -0,0 +1,93 @@
+
+resource "vsphere_tag_category" "category" {
+  name        = "openshift-${var.cluster_name}"
+  description = "Added by openshift-install do not remove"
```
maybe expand description, or add comments to explain what's this value

michaellevy101 replied:
The description is the tag description which should be applied on every resource, i should be "Created by the test-infra, do not delete manually" but as you understand i just copied it from the Openshift installer and not ready for a review (just asked to publish the code to provide a picture of the changes)..... I will document everything properly

osherdp commented on terraform_files/vsphere/main.tf:
```diff
@@ -0,0 +1,93 @@
+}
+
+resource "vsphere_folder" "folder" {
+  count = 1
```
```suggestion
  count         = 1
```

osherdp commented on terraform_files/vsphere/main.tf:
```diff
@@ -0,0 +1,93 @@
+  guest_id             = "coreos64Guest"
+  folder               = var.cluster_name
+  enable_disk_uuid     = "true"
+  wait_for_guest_net_timeout  = "0"
```
```suggestion
  wait_for_guest_net_timeout  = 0
```

michaellevy101 replied:
?

osherdp reviewed (changes requested):
well done for all the work
just a couple of fixes and we're good to go

nshidlin commented on discovery-infra/tests/base_test.py:
```diff
@@ -34,29 +38,105 @@
+            nat.remove_nat_rules()
+
+    @pytest.fixture
+    def cluster(self, api_client, request, proxy_server, prepare_network, cluster_configuration):
```
While `prepare_network` is the last fixture in the chain it says nothing about what the fixture does.

michaellevy101 replied:
prepare network is preparing the network.... in this case configuring the NAT. don't understand what do you mean by "says nothing about what the fixture does" can you please extend?

michaellevy101 commented:
@eliorerz @eliorerz @osherdp @YuviGold @nshidlin Thank you for your comments but this code isn't ready yet
I just asked to push the code to provide a picture of how to support multi-platform tests
I have addressed most of your comments

osherdp commented:
> @eliorerz @eliorerz @osherdp @YuviGold @nshidlin Thank you for your comments but this code isn't ready yet
> I just asked to push the code to provide a picture of how to support multi-platform tests
> I have addressed most of your comments

No problem @michaellevy101 
next time please add [WIP] for the PR title, so people will know 
description also

And again, great work is being done here. Very nice :)

osherdp commented:
/test ?

eliorerz commented:
/test ?

eliorerz commented:
/test e2e-metal-assisted-none e2e-metal-assisted-ipv6 e2e-metal-assisted
